Add content for how-to/download-new-upstream-version

dviererbe opened the pull request:
This pull requests adds the article for how to download a new upstream version.

Is related to #19

@s-makin and @keirthana you should be able to commit to this branch and merge it; feel free to make changes :)

dviererbe committed:
Add content for how-to/download-new-upstream-version

s-makin commented on docs/how-to/download-new-upstream-version.rst:
```diff
@@ -7,3 +7,63 @@ Download a new Upstream Version
 
     Work in progress!
+
+Once in a while you need to download a new upstream release or check if a newer 
```
Have we got a page that includes a definition what upstream is?

s-makin commented on docs/how-to/download-new-upstream-version.rst:
```diff
@@ -7,3 +7,63 @@ Download a new Upstream Version
 
     Work in progress!
+
+Once in a while you need to download a new upstream release or check if a newer 
```
Have we got a page that includes a definition what upstream is?

dviererbe replied:
I intend to add this to the not yet existing glossary and there is an explaination article planed that should explain the upstream and downstream concept (see https://github.com/canonical/ubuntu-packaging-guide/issues/19#issue-1789221296).

s-makin replied:
Sounds good to me :)

s-makin commented on docs/how-to/download-new-upstream-version.rst:
```diff
@@ -7,3 +7,63 @@ Download a new Upstream Version
+    ``README.debian`` file (if available) that tell you how to proceed.
+
+.. caution::
+    Only if there is no automatic download mechanism and you can't find any
+    download instructions, you should download any upstream file manually.
```
```suggestion
    You should only download any upstream file(s) manually if there is no automatic download mechanism and you can't find any download instructions. 
```

s-makin commented on docs/how-to/download-new-upstream-version.rst:
```diff
@@ -7,3 +7,63 @@ Download a new Upstream Version
+    Only if there is no automatic download mechanism and you can't find any
+    download instructions, you should download any upstream file manually.
+
+    Remember that a package may get distributed to hundred thousands of users.
```
```suggestion
    Remember that a package may get distributed to hundreds of thousands of users.
```

s-makin commented on docs/how-to/download-new-upstream-version.rst:
```diff
@@ -7,3 +7,63 @@ Download a new Upstream Version
+    wrong URL or copy a URL only partially, etc.
+
+If you just want to check if a new update is available, but you don't want to 
+download anything you can run the following command from the root of the 
```
```suggestion
download anything, you can run the following command from the root of the 
```

s-makin commented on docs/how-to/download-new-upstream-version.rst:
```diff
@@ -7,3 +7,63 @@ Download a new Upstream Version
+
+    uscan --safe
+
+.. note::
```
I think this is probably ok as just normal text (doesn't need to be a called-out note)

keirthana replied:
+1

s-makin commented on docs/how-to/download-new-upstream-version.rst:
```diff
@@ -7,3 +7,63 @@ Download a new Upstream Version
+:manpage:`uscan(1)` found)::
+
+    uscan --safe --verbose 
+
```
Might be worth adding a subheading on this line here, it seems sort of separate enough from the previous content.

s-makin commented on docs/how-to/download-new-upstream-version.rst:
```diff
@@ -7,3 +7,63 @@ Download a new Upstream Version
+
+    uscan --force-download
+
+.. note::
```
I think this is also probably ok as just regular text. It's a perfectly sensible explanation of what the previous command is doing :)

keirthana replied:
+1

s-makin commented on docs/how-to/download-new-upstream-version.rst:
```diff
@@ -7,3 +7,63 @@ Download a new Upstream Version
+    determine the name and the lastest upstream version of the source package.
+
+.. note::
+    If the upstream project has a signing key it should be stored in the source
```
Do we have any description/explanation of what a signing key is?

s-makin commented on docs/how-to/download-new-upstream-version.rst:
```diff
@@ -7,3 +7,63 @@ Download a new Upstream Version
+Further Information
+-------------------
+
+- man page -- :manpage:`uscan(1)`
```
```suggestion
- Manual page -- :manpage:`uscan(1)`
```

s-makin replied:
(otherwise, I'd have manpage as one word - I leave the decision to you!)

s-makin commented on docs/how-to/download-new-upstream-version.rst:
```diff
@@ -7,3 +7,63 @@ Download a new Upstream Version
+-------------------
+
+- man page -- :manpage:`uscan(1)`
+- debian wiki -- `debian/watch <https://wiki.debian.org/debian/watch>`_
```
```suggestion
- Debian wiki -- `debian/watch <https://wiki.debian.org/debian/watch>`_
```

s-makin commented on docs/how-to/download-new-upstream-version.rst:
```diff
@@ -7,3 +7,63 @@ Download a new Upstream Version
+
+- man page -- :manpage:`uscan(1)`
+- debian wiki -- `debian/watch <https://wiki.debian.org/debian/watch>`_
+- debian policy ``4.6.2.0`` -- `Upstream source location: debian/watch <https://www.debian.org/doc/debian-policy/ch-source.html#upstream-source-location-debian-watch>`_
```
```suggestion
- Debian policy ``4.6.2.0`` -- `Upstream source location: debian/watch <https://www.debian.org/doc/debian-policy/ch-source.html#upstream-source-location-debian-watch>`_
```

s-makin reviewed (commented):
Overall looks pretty good :) just a few small nits and suggestions. Thanks for writing it!

s-makin commented on docs/how-to/download-new-upstream-version.rst:
```diff
@@ -7,3 +7,63 @@ Download a new Upstream Version
+
+.. note::
+    If the source package does not contain a ``watch`` file, there may be
+    an explaination and instructions in the ``README.source`` or 
```
```suggestion
    an explanation and instructions in the ``README.source`` or 
```

s-makin committed:
Minor polishing

Sp/gr nits

keirthana commented on docs/how-to/download-new-upstream-version.rst:
```diff
@@ -7,3 +7,62 @@ Download a new Upstream Version
+    ``README.debian`` file (if available) that tell you how to proceed.
+
+.. caution::
+    You should only download any upstream file(s) manually if there is no automatic download mechanism and you can't find any download instructions. 
```
```suggestion
    You should download upstream file(s) manually only if there is no automatic download mechanism and you can't find any download instructions. 
```
Small nit, I just feel this reads better, WDYT?

keirthana commented on docs/how-to/download-new-upstream-version.rst:
```diff
@@ -7,3 +7,62 @@ Download a new Upstream Version
+    determine the name and the lastest upstream version of the source package.
+
+.. note::
+    If the upstream project has a signing key it should be stored in the source
```
> it should be stored in the source

It may be me but to a beginner's eyes, this is a bit ambiguous. Is there an action for the user in there or is this just a convention for the upstream project and we asking the user to look for the signing key. I definitely think a brief explanation for the signing key could help, like Sally pointed out.

dviererbe commented:
1. I also noticed that there is no example or mention of just running `uscan` without any parameter
2. Also my memory is playing tricks on me; the manpage url is not set in the `docs/conf.py`; therefore the `:manpage:` links do not work. I thought that I already configured this :/

dviererbe commented:
I just rebased this branch on the overhaul branch to get the commit (58d13fb), so that the `:manpage:` refs work.

dviererbe committed:
Restructered how-to/download-new-upstream-version

Mostly the same content, just restructured/enhanced:
- incorporated wording/structure feedback from Sally and Keirthana
- added subheadings
- demoted notes to paragraph
- added :term:, :file: sphinx roles
- replaced code-block with `.. code:: bash` syntax
- added reference to get-package-source to uscan forced download

Co-authored-by: Sally <sally.makin@canonical.com>
Co-authored-by: Keirthana T S <keirthana.ts@canonical.com>

dviererbe committed:
Add back missing WIP banner to how-to/download-new-upstream-version

keirthana reviewed (approved):
LGTM!

vpa1977 commented on docs/how-to/download-new-upstream-version.rst:
```diff
@@ -1,9 +1,97 @@
+
+    uscan --safe
+
+Force Download of latest :term:`Upstream` Version
```
maybe worth noting that we can do `--download-current-version` to get orig.tar.gz (when working outside the git launchpad git tree and git-ubuntu export-orig is not available)  for example when working with Debian upstream.
We can also download arbitrary version through `--download-version` option which is useful when we need to pack some arbitrary release.

dviererbe replied:
I left them out, because I never used these flags before, but now that you have mention them independently I will add them to the how-to.

dviererbe committed:
Explain uscan flags --download-version and --download-current-version

s-makin commented:
> I just rebased this branch on the overhaul branch to get the commit ([58d13fb](https://github.com/canonical/ubuntu-packaging-guide/commit/58d13fb25b7bb462518b77ee2841cac058423cc4)), so that the `:manpage:` refs work.

The manpage link for uscan doesn't seem to be working. When I go via the manpages, all of the links work except for mantic - there doesn't seem to be a mantic page for uscan, the page keeps timing out.

s-makin commented:
> > I just rebased this branch on the overhaul branch to get the commit ([58d13fb](https://github.com/canonical/ubuntu-packaging-guide/commit/58d13fb25b7bb462518b77ee2841cac058423cc4)), so that the `:manpage:` refs work.
> 
> The manpage link for uscan doesn't seem to be working. When I go via the manpages, all of the links work except for mantic - there doesn't seem to be a mantic page for uscan, the page keeps timing out.

Same for the links to git manpage. 

It may be better for us to configure the manpages link to make them version agnostic by doing: `manpages.ubuntu.com/manpages/<package-name>.html` - this currently redirects to the Lunar version of the page by default, but means we wouldn't have to update it manually each time it gets out of date (reduces ongoing maintenance burden)

s-makin commented on docs/how-to/download-new-upstream-version.rst:
```diff
@@ -1,9 +1,139 @@
+
+    uscan --verbose
+
+Check for new :term:`Upstream` version (no download)
```
I'd suggest we move this to be the first sub-heading. Any user may want to check their version before downloading, so I think that would be a more logical ordering.

dviererbe replied:
You don't have to run `uscan --safe` to run `uscan`. The `--safe` flag basically disables downloading anything. Both are for 2 different use cases.

You would run `uscan`, e.g.:
- as the package maintainer to download and package the new release

You would run `uscan --safe`, e.g.:
- during a [Main Inclusion Review](https://github.com/canonical/ubuntu-mir) just to check if the `debian/watch` file works
- [while fixing a bug](https://github.com/canonical/ubuntu-maintainers-handbook/blob/f788e02eccb4a7cf9696585e4e7634d8d2735ed7/PackageFixing.md#was-it-fixed-upstream) just to make sure that there is no newer version that may have fixed the bug already

dviererbe replied:
@s-makin With that said; do you still think that the order should be moved?

s-makin replied:
My impression is that checking the debian/watch file works or fixing a bug might be easier or at least more common steps that a packager would need to know before they become a package maintainer (although I could be wrong there!). If that's correct, then I'd move the `uscan --safe` section. Otherwise I'd leave it as-is - I'll bow to your greater knowledge on this one!

dviererbe replied:
Actually I don't know which use case is more common. I will ask around in `Foundations`

s-makin commented on docs/how-to/download-new-upstream-version.rst:
```diff
@@ -1,9 +1,139 @@
+    :file:`debian/README.debian` file (if available) that tell you how to proceed.
+
+.. caution::
+    You should download :term:`Upstream` file(s) manually only if there is no automatic
```
As a general suggestion, I'd say we should only define terms via the glossary the first time they appear in the page, the same way we would with an unexpanded acronym. Otherwise it's much more work than it should be to write an article.

dviererbe replied:
My thought was – if we reference the terms always throughout the article – readers which only read parts of an article have a much better time, because they don't have to search for the first appearance.

If you still think that this is not worth it, than I will only include the first appearance.

s-makin replied:
I think it's generally fair to assume (at least, it's my working principle, for what it's worth) that someone who is new would be more likely to read the whole thing, while those who skip ahead to the bits they already know they want probably won't need the glossary definitions so much. There's definitely no harm in leaving them in if you want to, but since you've got a lot of writing ahead I wanted to share that perspective to save you potentially a lot of time and energy.

dviererbe replied:
I will continue with the "always reference" approach for now, because I actually use the search and replace feature of my editor; therefore it is not much more work, but I keep this advice in the back of my mind and look how it goes.

s-makin commented on docs/how-to/download-new-upstream-version.rst:
```diff
@@ -1,9 +1,139 @@
 
     Work in progress!
+
+Once in a while you need to download a new :term:`Upstream` release or check if a newer
```
```suggestion
Once in a while you may need to download a new :term:`Upstream` release or check if a newer
```

s-makin replied:
I would also, as a user (particularly a new user) ask "why/under what circumstances?"

dviererbe replied:
There are many circumstances why you actually want to do this, but I can list some example circumstances like the ones I listed above:

> You would run uscan, e.g.:
>
> - as the package maintainer to download and package the new release
>
> You would run uscan --safe, e.g.:
>  
> - during a [Main Inclusion Review](https://github.com/canonical/ubuntu-mir) just to check if the debian/watch file works
> - [while fixing a bug](https://github.com/canonical/ubuntu-maintainers-handbook/blob/f788e02eccb4a7cf9696585e4e7634d8d2735ed7/PackageFixing.md#was-it-fixed-upstream) just to make sure that there is no newer version that may have fixed the bug already

s-makin replied:
I think using one example (max two) is fine, and I'd say probably whichever you think is most relevant to someone who might be approaching this task for the first time. We don't want to turn this into an explanation page, but still, for a newer developer it might not be obvious why this task is useful.

s-makin commented on docs/how-to/download-new-upstream-version.rst:
```diff
@@ -1,9 +1,139 @@
+    an explanation and instructions in the :file:`debain/README.source` or
+    :file:`debian/README.debian` file (if available) that tell you how to proceed.
+
+.. caution::
```
This caution box is quite large when rendered on the page, and it seems to be mostly relevant to the download section. Can we incorporate this cautionary content into the instructions in that section?

dviererbe replied:
@s-makin What do you think of converting the caution box into a section "Best practices"

s-makin replied:
I think that's a great idea!

s-makin commented on docs/how-to/download-new-upstream-version.rst:
```diff
@@ -1,9 +1,139 @@
+----------------------------------------------------
+
+Running :manpage:`uscan(1)` from the :term:`Root` of the :term:`Source Tree` will
+check if a newer :term:`Upstream` version exits and downloads it:
```
```suggestion
check if a newer :term:`Upstream` version exists and downloads it:
```

s-makin commented on docs/how-to/download-new-upstream-version.rst:
```diff
@@ -1,9 +1,139 @@
+    
+    There are special cases where they do not work for technical reasons.
+
+.. note::
```
This note appears twice - is that intentional?

dviererbe replied:
Yes, this was intentional for readers that read the article only partially. I also don't think that it is elegant.

Do you have a more elegant solution?

s-makin replied:
Hmmm...What if the header was something like "Downloading upstream versions", with the note contents underneath that heading, and then the two current headings become subheadings beneath that? I'm not sure that the note would make so much sense then, but at least it would remove the duplication. Might need an additional sentence or two to introduce the difference? (sadly, that's the best I have!)

s-makin reviewed (commented):
Just a couple of comments/suggestions, but otherwise looks good :)

dviererbe committed:
typo: 'exits' -> 'exists'

Co-authored-by: Sally <sally.makin@canonical.com>

dviererbe committed:
chore: minor wording improvement

Co-authored-by: Sally <sally.makin@canonical.com>

dviererbe commented:
> > > I just rebased this branch on the overhaul branch to get the commit ([58d13fb](https://github.com/canonical/ubuntu-packaging-guide/commit/58d13fb25b7bb462518b77ee2841cac058423cc4)), so that the `:manpage:` refs work.
> > 
> > 
> > The manpage link for uscan doesn't seem to be working. When I go via the manpages, all of the links work except for mantic - there doesn't seem to be a mantic page for uscan, the page keeps timing out.
> 
> Same for the links to git manpage.
> 
> It may be better for us to configure the manpages link to make them version agnostic by doing: `manpages.ubuntu.com/manpages/<package-name>.html` - this currently redirects to the Lunar version of the page by default, but means we wouldn't have to update it manually each time it gets out of date (reduces ongoing maintenance burden)

That is weird, because the links work for me. I get redirected from `mantic` to `lunar`.

How would `manpages.ubuntu.com/manpages/<package-name>.html` work for linking e.g. `uscan`? `uscan` is part of the [`devscripts`](https://manpages.ubuntu.com/manpages/lunar/en/man1/devscripts.1.html) package which itself has multiple tools.

dviererbe committed:
chore: Add exaple use cases for uscan

dviererbe committed:
chore: convert caution box to section

The caution box in the article how-to/download-new-upstream-version that
instructs the user to avoid downloading Upstream Files manually gets
converted into its own section ("Best practices").

dviererbe committed:
fix(how-to/download-new-upstream-version): remove term refs from headings

dviererbe committed:
chore(how-to/download-new-upstream-version): change code to code-block

dviererbe committed:
chore(how-to/download-new-upstream-version): remove overused term refs
